webhooks/services.py

```diff
@@ -1,14 +1,137 @@
+                .replace("'", '"')
+                .replace("False", json.dumps(False))
+                .replace("True", json.dumps(True))
+                .replace("None", json.dumps(None))
+            )
+        except JSONDecodeError:
+            # Don't raise an error here. It's probably more valuable to
+            # Return the data anyway than breaking the retrieval
+            return {"invalid JSON data": data}
+
+    def _to_json(self, data: List[bytes]) -> List[dict]:
+        data = [self._deserialise(val) for val in data]
+        try:
+            data = [json.loads(d) for d in data]
+        except JSONDecodeError:
+            # A multiline string has occured that has been split in
+            # an invalid way wrt JSON format
+
+            # Remove multiple spaces which occur when error msg is split into multiple lines
+            # turn off black formatter here because it replaces the regex
+            # fmt: off
+            data = [re.sub('\s+',' ', d) for d in data]
+            # Remove [..\" \"..] from the multiline error message because that is invalid json
+            data = [d.replace('\" \"', ' ') for d in data]
+            # fmt: on
+            data = [json.loads(d) for d in data]
+        return data
+
+    def _to_item(self, topic: str, data: List[dict]) -> List[dict]:
+        # Transform the data to the appropriate model
+        data = self._to_json(data=data)
+        if self._is_proof(topic):
+            data = [self._proof_hook_versioned(d) for d in data]
+        elif self._is_credential(topic):
+            data = [self._credential_hook_versioned(d) for d in data]
+        elif topic == "connections":
+            data = [self._version_connections(d) for d in data]
+        elif topic == "basicmessages":
+            data = [BasicMessagesHook(**d) for d in data]
+        return data
+
+    def _to_topic_item(self, topic: str, data: List[bytes]) -> List[TopicItem]:
+        data = self._to_item(topic=topic, data=data)
+        return (
+            [
+                TopicItem(
+                    topic=topic,
+                    wallet_id=payload.wallet_id,
+                    origin=payload.origin,
+                    payload=payload.dict(),
+                )
+                for payload in data
+            ]
+            if data
+            else []
+        )
+
+    async def add_topic_entry(self, topic: str, hook: bytes) -> List[TopicItem]:
         return await self._redis.sadd(topic, hook)
 
-    async def get_all_by_topic(self, topic: str) -> str:
-        return await self._redis.smembers(topic)
+    async def get_all_by_topic(self, topic: str) -> List[TopicItem]:
+        # Get the data from redis queue
+        data = await self._redis.smembers(topic)
+        # Transform it to TopicItems
+        items = self._to_topic_item(topic=topic, data=data)
+        for d in items:
+            del d.payload["wallet_id"]
+            del d.payload["origin"]
```

Comment:
If we don't add these keys to the payload itself, but rather create an object like this:

```
{
   "wallet_id": "xx",
   "origin": "xxx",
   "data": {}
}
```

Then we don't have to do the deleting of the properties before returing (can just return data) and we also avoid the possiblity of the attributes magically clashing. (the wallet record has a property named `wallet_id`)

Reply:
ah i remember now why I did that. Because the dtaa living in redis only has a single key and the wallet id and origin attrs get added when the webhook is received and then stored to redis under a single key. So when retrieving the data from redis we wnt that piece of informtion to be there but not in the "payload" that represents some model from aca-py or cloudapi. Hence, it's being deleted. Not sure how to get around this otherwise.

Reply:
found a way nontheless i think. Will push in a little

Reply:
They can just get ditched when transforming into models